api/apps/api/src/modules/cost-surface/cost-surface.service.ts

```diff
@@ -118,4 +122,33 @@ export class CostSurfaceService {
 
     return right(costSurface);
   }
+
+  async getCostSurfaceRange(
+    costSurfaceId: string,
+    projectId: string,
+    userId: string,
+  ): Promise<
+    Either<
+      typeof costSurfaceNotFoundForProject | typeof projectNotEditable,
+      CostRange
+    >
+  > {
+    if (
+      !(await this.projectAclService.canEditCostSurfaceInProject(
```

Comment:
well, actually.

The trouble with the kind of logic we have here for ACLs (I've been yelling at clouds about this, just not with you... yet 😅) is that in practice, all we need to get in the endpoint URL is a `scenarioId`, but we include a `projectId` for reasons (URL design blah blah), but then, I'm not really thrilled about the way we use these.

If a malicious user somehow knows the UUID of a cost surface (or any other thing, such as a feature, the overall logic is the same) in a project they don't have access to, and request a cost surface (range, in this case) or tile, or feature, or whatever, supplying the UUID of the thing they want to see, and the projectId of any project they have access to... in your implementation here they won't have any luck because the `findOne()` below looks for a specific cost surface in a specific project, so they will receive a 404. But my point is, this is slightly brittle. If someone comes in months from now and makes some changes without full context, for example decide to simplify the `where` for whatever reason and only search for the `costSurfaceId` because hey, that's unique anyway, then the malicious user above will win (tm).

So I'm happy to keep this as is, because it works and it's a low-risk scenario in any case, but in practice my recommendation for these kinds of checks would be to outright  ignore the `projectId` supplied by the user, and instead look up the `projectId` from the supplied `scenarioId`, _and then_ check if the user can view the project.

I would not even extract the `projectId` from the request, nor pass it on from the controller handler to `CostSurfaceService.getCostSurfaceRange()` - just `scenarioId` and `userId`. Just keep the `projectId` in the URL to make it look kind of logical-looking but then ignore it. I'd even go as far as removing the `/projects/:projectId` part of the URL, but we do have this kind of parent/child URL design in several places, so I'd just leave this as is.